hypertrace-gradle-docker-plugin/src/main/java/org/hypertrace/gradle/docker/DockerImage.java

```diff
@@ -41,22 +42,21 @@ public DockerImage(
     this.name = name;
     this.extension = extension;
     this.buildArgs = objectFactory.mapProperty(String.class, String.class);
-    this.imageName = objectFactory.property(String.class)
-                                  .convention(name);
+    this.imageName = objectFactory.property(String.class).convention(name);
+    this.platforms =
+        objectFactory.listProperty(String.class).convention(List.of("linux/amd64"));
```

Comment:
I couldn't build docker image locally with multiple platforms passed like `--platform linux/amd64,linux/arm64`. But if I pass in either of those it works fine. Buildkit comes with docker desktop and tried some online references to make it work but didn't work for me (mine is M1 mac)

For CI I found action reference at https://nielscautaerts.xyz/making-dockerfiles-architecture-independent.html . I would want to default this property to both arm and amd, but we need to figure out how to get docker build locally first by passing multiple args.

Reply:
Error I get & buildx config
```
➜  java-21 git:(multi-arch) pwd
/Users/kishanadapa/dev/hypertrace/docker-java-base/java-21
➜  java-21 git:(multi-arch) ls -l
total 16
-rw-r--r--  1 kishanadapa  staff  3006 Feb  6 15:41 Dockerfile
-rw-r--r--  1 kishanadapa  staff   342 Feb  5 15:04 build.gradle.kts
➜  java-21 git:(multi-arch) docker buildx build . -t test-21-multi-arch:1 --platform linux/amd64,linux/arm64 --progress=plain --build-arg JAVA_VERSION=21.0.1-21.30.15
ERROR: Multiple platforms feature is currently not supported for docker driver. Please switch to a different driver (eg. "docker buildx create --use")
➜  java-21 git:(multi-arch) docker buildx ls
NAME/NODE       DRIVER/ENDPOINT STATUS  BUILDKIT             PLATFORMS
default         docker
  default       default         running v0.11.6+0a15675913b7 linux/arm64, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6
desktop-linux * docker
  desktop-linux desktop-linux   running v0.11.6+0a15675913b7 linux/arm64, linux/riscv64, linux/ppc64le, linux/s390x, linux/386, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6
```

Reply:
ah, looks like this aligns with the comment I mentioned elsewhere. Given we actually use buildx directly in CI, I don't think we need to get the gradle build to support it (as it doesn't look like it will be able to). I would aim for (it sounds like maybe this is already your goal?):
- locally, it dockerBuildImages (Via gradle) should build for the current platform
- via buildx, the same generated dockerfile should be able to build multiplatform. If this can't be supported, we should be able to trivially wrap it such that we can create a shared action to build a multiplatform image out of the generated one.

Hope that makes sense

Reply:
> locally, dockerBuildImages (via gradle) should build for the current (local) platform

The problem with this is when people build custom images for debugging, it will be on arm but our machines are on x86. So we either want a multi arch image or else default to x86 for local.

> don't think we need to get the gradle build to support it (as it doesn't look like it will be able to)

Platforms is a argument in gradle plugin, it should support it. Need to figure out why local docker engine (irrespective of gradle) is not accepting comma delimited platforms. In few places I read it should be supported, may be I am missing something

> via buildx, the same generated dockerfile should be able to build multiplatform.

Yes on CI, it is not a problem. Even if `dockerBuildImages` defaults to only x86 it is okay, we can pass in platforms arg to docker build publish action

Reply:
> or else default to x86 for local.

Multiplatform is fine, but it shouldn't default to x86. If we can't get it to work with multiplatform it should detect and use the local arch.

> Platforms is a argument in gradle plugin, it should support it.

It supports a platform arg, but it doesn't support buildx and building multiplatform images.

> Need to figure out why local docker engine (irrespective of gradle) is not accepting comma delimited platforms, but works with either of individual. In few places I read it should be supported, may be I am missing something.

You need to set up a builder first. See https://docs.docker.com/build/building/multi-platform/#building-multi-platform-images

```
docker buildx create --platform linux/arm64,linux/amd64 --name multiplatform-builder

docker buildx build . -t test-21-multi-arch:1 --builder multiplatform-builder --platform linux/amd64,linux/arm64 --build-arg JAVA_VERSION=21.0.1-21.30.15 --load
```

Succesfully created a local multiplatform image for me (have to enable containerd in docker engine to load a multiplatform image though)
![image](https://github.com/hypertrace/hypertrace-gradle-docker-plugins/assets/45047841/42fe5de0-15f9-41d1-9260-a8648c636a3e)

Reply:
Actually just the flip of `containerd` in settings (I had to update to newer docker desktop) seems to have done the trick. I removed the newly added builder, did a system prune -a to clear all caches & images and built it again removing all the buildx params
```sh
➜  java-21 git:(multi-arch) docker buildx ls
NAME/NODE       DRIVER/ENDPOINT STATUS  BUILDKIT PLATFORMS
default         docker
  default       default         running v0.12.5  linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64
desktop-linux * docker
  desktop-linux desktop-linux   running v0.12.5  linux/arm64, linux/amd64, linux/amd64/v2, linux/riscv64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64
➜  java-21 git:(multi-arch) docker build . -t another-test-21-multi-arch:1 --platform linux/amd64,linux/arm64 --build-arg JAVA_VERSION=21.0.1-21.30.15 --load
<worked fine>
```

> It supports a platform arg, but it doesn't support buildx and building multiplatform images.

So the plugin in the end passes these args to docker daemon right, given the docker desktop itself is defaulting to buildx (on mac atleast), shouldn't it be enough to just pass platform arg? We should ask all to turn on this containerd flag though. If we verify that just that alone works on a x86 mac and (another) M1 mac, we should be good I think

Reply:
That didn't seem to work on first try using snapshot build
```
Error during callback
com.github.dockerjava.api.exception.BadRequestException: Status 400: {"message":"\"amd64,linux\" is an invalid component of \"linux/amd64,linux/arm64\": platform specifier component must match \"^[A-Za-z0-9_-]+$\": invalid argument"}

        at com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:237)
        at com.github.dockerjava.core.DefaultInvocationBuilder.lambda$executeAndStream$1(DefaultInvocationBuilder.java:269)
        at java.base@11.0.11/java.lang.Thread.run(Thread.java:829)
```

Reply:
Okay so I get it now, `buildx` is a CLI plugin and the docker-java talks to the Docker Engine API, which isn't aware of all the magic this plugin does hence doesn't work

Reply:
> Multiplatform is fine, but it shouldn't default to x86. If we can't get it to work with multiplatform it should detect and use the local arch.

I confirmed that default behaviour is to use local arch (used base image to be the one we just built on this thread). That means we can close out this PR as it is not needed and merge https://github.com/hypertrace/docker-java-base/pull/28

Only issue is when people need to generate custom image for x86_64 to deploy on servers we need them to cd into Dockerfile context and build the image